mobile-app/app/screens/AppNavigator/screens/Dex/CompositeSwap/CompositeSwapScreen.tsx

```diff
@@ -0,0 +1,759 @@
+  // Use Sets to reduce checks if item is unique
+  const visitedNodes = new Set<string>([])
+  const nodesToVisit = new Set<string>([tokenFrom])
+
+  while (nodesToVisit.size !== 0) {
+    const [token] = nodesToVisit // first item in a set
+    const adjacentNodes = getAdjacentVertices(tokenFrom, graph)
+
+    if (adjacentNodes !== undefined) {
+      adjacentNodes.forEach(node => {
+        if (!reachableNodeIds.has(node)) {
+          reachableNodes.push(node)
+        }
+
+        // If the token hasn't been visited, flag for visit.
+        if (!visitedNodes.has(node)) {
+          nodesToVisit.add(node)
+        }
+        reachableNodeIds.add(node)
+      })
+    }
+
+    visitedNodes.add(token)
+    nodesToVisit.delete(token)
+  }
+
+  return reachableNodes.reduce((tokens: BottomSheetToken[], node: string): BottomSheetToken[] => {
+    const token = allTokens.find(token => token.id === node)
+    if (token !== undefined && node !== tokenFrom) {
+      return [
+        ...tokens, {
+          tokenId: token.id,
+          available: new BigNumber(token.reserve),
+          token: {
+            name: '', // Not available in API
+            displaySymbol: token.displaySymbol
+          }
+
+        }
+      ]
+    }
+
+    return tokens
+  }, [])
+}
+
+interface GraphProps { pairId: string, a: string, b: string }
+/**
+ * To find all the adjacent vertices / neighbors of a node in a graph
+ * @param startNode
+ * @param graph
+ */
+function getAdjacentVertices (startNode: string, graph: GraphProps[]): string[] {
+  return (
+    graph.reduce((vertices: string[], vertex): string[] => {
+      if (vertex.a === startNode && vertex.b !== startNode) {
+        return [...vertices, vertex.b]
+      } else if (vertex.b === startNode && vertex.a !== startNode) {
+        return [...vertices, vertex.a]
+      }
+
+      return vertices
+    }, [])
+  )
+}
+
+/**
+ * This uses a modified Breadth First Search to find the first path found (by distance)
+ * @param graph
+ * @param start
+ * @param target
+ */
+function findPath (graph: GraphProps[], origin: string, target: string): string[] {
+  let isPathFound = false
+  let nodesToVisit = [origin]
+  const visitedNodes = new Set<string>([]) // track visited nodes in a set
+  let currentDistance = 0 // track distance from origin to target
+  const path: string[] = [] // store the first path found by token
+
+  bfs({
+    start: {
+      value: origin,
+      edges: getAdjacentVertices(origin, graph)
+    },
+    target: target
+  })
+
+  function bfs ({ start, target }: { start: { value: string, edges: string[]}, target: string}): void {
+    if (!isPathFound && (start.edges.length !== 0 || start.value === target)) { // not yet a dead-end
+      path[currentDistance] = start.value // store possible path to to target
+    }
+
+    if (start.value === target) {
+      isPathFound = true
+    }
+
+    visitedNodes.add(start.value)
+
+    while (nodesToVisit.length > 0) {
```

Comment:
Need to optimize with early return, I believe the search keep going even path found now.

let say your found path = [BTC, DFI, ETH]

The logic will keep going when more DEX pair available 

n: DFI
n + 1: TSLA, AMZN, AAPL.... // should end here once ETH found
n + 2: DUSD
n + 3: if any XXX-DUSD found but not XXX-DFI

Reply:
Optimized to stop once a path is found and moved the `findPath` logic to another file as preparation for e2e (working on it now)